urbanairship-react-native/js/UAEventEmitter.ts

```diff
@@ -2,42 +2,105 @@
 
 'use strict';
 
-import { NativeModules, NativeEventEmitter, EmitterSubscription, Platform } from "react-native";
+import { NativeEventEmitter, Platform, AppRegistry, NativeModules } from "react-native";
 
 /**
  * @hidden
  */
-const UrbanAirshipModule = NativeModules.UrbanAirshipReactModule;
+export type DispatchEventsCallback = (source: (eventType: string) => Promise<any>) => Promise<any>;
 
 /**
- * Custom native event emitter with additional Android behavior
- *
  * @hidden
  */
-export class UAEventEmitter extends NativeEventEmitter {
-  constructor() {
-    super(UrbanAirshipModule);
+export abstract class AirshipEventBridge {
+  dispatchEventsCallback: DispatchEventsCallback;
+
+  constructor(dispatchEventsCallback: DispatchEventsCallback) {
+    this.dispatchEventsCallback = dispatchEventsCallback;
   }
 
-  addListener(eventType: string, listener: (...args: any[]) => any, context?: Object): EmitterSubscription {
+  abstract notifyAirshipListenerAdded(eventListener: string): void;
+}
+
+class DefaultAirshipEventBridge extends AirshipEventBridge {
+  nativeModule = NativeModules.UrbanAirshipReactModule;
+  eventEmitter = new NativeEventEmitter(this.nativeModule);
+
+  constructor(dispatchEventsCallback: DispatchEventsCallback) {
+    super(dispatchEventsCallback);
+
     if (Platform.OS === 'android') {
-      UrbanAirshipModule.addAndroidListener(eventType);
+      AppRegistry.registerHeadlessTask('AirshipAndroidBackgroundEventTask', () => {
+        return () => dispatchEventsCallback(this.nativeModule.takePendingBackgroundEvents);
+      });
+
+      this.eventEmitter.addListener("com.urbanairship.onPendingForegroundEvent", async () => {
+        return dispatchEventsCallback(this.nativeModule.takePendingForegroundEvents);
+      });
+    } else if (Platform.OS === 'ios') {
+      this.eventEmitter.addListener("com.urbanairship.onPendingEvent", async () => {
+        return dispatchEventsCallback(this.nativeModule.takePendingEvents);
+      });
     }
-    return super.addListener(eventType, listener, context);
   }
 
-  removeAllListeners(eventType: string) {
-    if (Platform.OS === 'android') {
-        UrbanAirshipModule.removeAndroidListeners(this.listeners(eventType).length);
-    }
+  notifyAirshipListenerAdded(eventType: string): void {
+    this.nativeModule.onAirshipListenerAdded(eventType);
+  }
+}
+
+/**
+ * SDK event emitter.
+ *
+ * @hidden
+ */
+export class UAEventEmitter {
+  airshipEventBridge: AirshipEventBridge;
+  listeners: Map<string, Array<(...args: any[]) => any>>;
```

Comment:
I kind of feel like if we're moving away from the underlying event emitter we might be able to get away with a more explicit typing for our listener functions than just `(...args: any[]) => any`, but maybe we can save that for another time

Reply:
The task provider does not seem to take an async callback